### ISSUE TITLE
test(model): isolate Metal P4 input normalization against serial fixture

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12853]
Parent: #12349
Coordinates with: #12843
Coordinates with: #12837

## Parent context
#12349 requires actual batched Metal verification with exact serial logits/state. Its strict M3 Pro P4 witness executed and failed; normalization has not yet been isolated.

## Current state
The strict synthetic Q4K/Q6K/Q8 four-layer fixture first reports layer0 conv[0] bf7e14eb versus bf7e14f0. The conv window stores raw mixed QKV values, so the mismatch is upstream of recurrence. Serial normalization uses host normCfg; P4 uses graph.RMSNorm. Projection arithmetic also differs, and the earliest differing operation remains unobserved.

## For / Problem / Today / Better because
For native Qwen MTP users, isolate the first actual numerical divergence before changing kernels. Today the end-to-end failure cannot attribute norm versus quantization/projection. A physical norm-only diagnostic on the exact fixture's real draft input resolves that uncertainty without changing supported shapes or tolerances.

## Why this is next
Layer0 normalization is the earliest differing operation in source; no projection or recurrence experiment should precede this observation.

## Working spine
Actual fixture draft embeddings [3,5,7,11], same scaleEmbedInPlace and layer0 norm weights -> existing graph.RMSNorm P4 -> terminal readback -> Float32bits comparison of token7 row with existing host normCfg -> captured first mismatch/equality and immutable input identity.

## Core through-line
Expose only the exact layer0 normalization boundary using the existing production fixture and graph operation, preserving every strict parent oracle and observing before changing arithmetic.

## Scope
```routing
lane: model
paths: ["internal/model/metal_mtp_verify_panel_darwin_test.go"]
expected_steps: 3
```
Add narrowly scoped normalization observability inside TestMetalMTPP4StrictSerialParity; preserve its strict logits/state/hidden-stage assertions and the frozen production downgrade patch. No new production helper or alternate model input.

## Scope class
Work unit: leaf
Single S1 test-only diagnostic in one existing file.

## Definition of done
- [ ] Exact fixture/input diagnostic source reviewed and frozen.
- [ ] Assigned physical run captures norm equality or first bit mismatch, with no skip and required graph completion.
- [ ] Result reconciled to parent and independently reviewed; no production or tolerance changes.

## Gold-plating boundary
No normalization fix, kernel, quantization/projection probe, scope exclusion, tolerance adjustment or performance claim. A mismatch is diagnostic evidence, not a reason to weaken parent equality. #12843 is a separate P1 admission/tolerance leaf, not a duplicate of this exact P4 norm diagnostic; #12837 CPU RoPE is only related context.

## Acceptance gate
Independent source review binds the exact four-row embedding panel and actual norm weights to the parent fixture. Required actual-device completed norm graph, one terminal readback, explicit exact bit result, input/weight hashes and terminal handles. Missing/skipped test earns no credit. Production unchanged.

## Done condition
One diagnostic is reviewed and executed in an assigned bounded window, with result captured and reconciled to #12349. If norm is equal, scope the next quant/projection diagnostic separately; if unequal, retain its first coordinate/bits before proposing any repair. Full parent parity remains open.

## Witness
env -u FAK_METAL_MPS CGO_ENABLED=1 GOMAXPROCS=2 go test -p=1 -timeout=60s ./internal/model -run '^TestMetalMTPP4StrictSerialParity$' -count=1 -json

The parent test currently fails strict downstream parity, so process failure must be classified by captured assertions. This diagnostic has its own explicit norm-stage equality/mismatch marker; a known downstream failure is not norm evidence.

## Operating envelope and budget
One approved invocation on Apple M3 Pro, native Metal, source-defined H256/I256/vocab512/4layer Q4K+Q6K+Q8 fixture, prefix32,P4, SDOT=false. Existing strict test costs two32token prefills+oneP4+fourserial+two continuation steps. Added norm probe is one4x256 input upload (4096bytes),256norm weights(1024bytes),oneRMSNorm encoder,onecommandbuffer/terminalwait/readback(4096bytes). Test timeout60s; outer compile ceiling300s. No HIL until exact candidate review and separate slot grant. Real model/performance qualification remains absent.

## Duplicate search
Open/closed searches for RMSNorm P4 and norm Metal parity inspected #12843 (P1 ordered attention/final norm admission) and #12007 (CPU HAL norm semantics); neither covers this actual P4-versus-serial first-divergence diagnostic.

## Problem frame
Centrality: Core
Portfolio tier: 2 serving
P1: preserved - exact state equality stays required.
P2: preserved - no speed claim from a diagnostic.
P3: preserved - one reversible test-only scope.
P4: advanced - actual route and input diagnostic.

## Work estimate
Estimate: 1 point; bounded one-file diagnostic, uncertain physical result.

## Overall completion contribution
Contribution: 1/4 points within #12349's original four-step contract; planning estimate, not percent completion or hardware qualification.

## Completion standard
experiment

## Closure binding
The explicit-path diagnostic commit cites this child and (fak model); parent #12349 remains open until its original strict acceptance and integration are satisfied.

**Comment by anthony-chaudhary** [comment]
Physical diagnostic executed once on Apple M3 Pro, native Metal, with reviewed synthetic P4 fixture and exact bit oracle. `TestMetalMTPP4StrictSerialParity` ran (0.20s; package 0.659s), exit 1, no skips. This is diagnostic RED, not parent parity success.

Layer0 token7 normalization differs before any projection: element0 Metal `-2.9536178 [c03d0813]`, host `-2.953618 [c03d0814]`. Input panel SHA256 `7f0dc47c86fdc84fe42f60e66a5edd2fb6f402202bcd9479c904ce08319b9a95`, weights `893a106828fbdb9521e1d868c985aab7ad2ae2f606edc55329265a5e7676006c`. One norm graph completed with one encoder and terminal readback. This isolates a real normalization mismatch; it does not yet attribute reduction versus rsqrt/product rounding or establish that its correction fixes every downstream difference.

Hidden-stage diagnostics both match immediately after Verify and Commit: 36 rows, 36 tokens, suffix `[3 5 7 11]`. The target transaction retains its suffix; later serial-oracle comparison remains a separate investigation. Original strict logits/GDN/KV/continuation failures remain.

Command: `env -u FAK_METAL_MPS CGO_ENABLED=1 GOMAXPROCS=2 go test -p=1 -timeout=60s ./internal/model -run '^TestMetalMTPP4StrictSerialParity$' -count=1 -json`.

Base `824ced654eb26382f7e637169dc40032a9604235`; test file SHA256 `fb5e62c18532d62fe75b9c8d7d1deffede50012114a9e042896ff6097bd02c95`; reviewed composite patch `7c32976de9e5209c3d4e55d16a4951fd5be0bc2904881e1c698880e889c2bd84`. Pre/post source manifest hashes identical `fb2645b377810e9b9d262c9dacc36dee80b363682fa318e1826ab581e4c70088`; no #12837 overlay. Unlanded diagnostic source; independent result review and governed landing remain pending. Parent #12349 stays open with exact equality unchanged.

**Comment by anthony-chaudhary** [comment]
Next bounded diagnostic within existing one-test scope: on the exact captured token7 input, calculate A sequential float32 sum and B the 256-leaf Metal reduction order; both retain CPU sqrt64 reciprocal/cast and output multiplication. Log sum/inverse/output bits or hashes and compare B with actual graph output. Preserve existing strict assertions, input/weights and native device requirements. This separates a reduction-order hypothesis from remaining rsqrt/output arithmetic differences; equality cannot establish a universal cause. No production/shader/tolerance changes. Fresh source review before any new HIL.

**Comment by anthony-chaudhary** [comment]
Required native M3 Pro diagnostic executed once (session59520 exit1; 1 named run, 0 skips). A sequential sum405e696b/inv41094660 retains prior host output mismatch. B 256-leaf tree sum405e696f/inv4109465f, using the SAME CPU sqrt64 reciprocal/cast, matches all256 actual graph output bits (SHA47abe4ae51eab1bc5d04fc3c7f94e964080da900bea8826766c98937880c491b). This reduction-order-only host hypothesis reproduces the captured token7 norm output; it does not prove universal rsqrt equivalence or sole cause of later drift. Original strict downstream assertions remain RED. Test07e5/source manifest a1e923 unchanged before/after. Receipt and raw/readable logs: /tmp/fak-12853-reduction-run/. CPU/HIL slot released; no retries or production changes. Next: independent evidence review and a bounded arithmetic contract decision.

**Comment by anthony-chaudhary** [comment]
Next one-file diagnostic layered on immutable #12889 shader53386: compare all4 norm rows, extend existing norm graph with only layer0 in_proj_qkv Q8 projection and terminal pack, then compare token7 against original serial mulGroup quartet once. Require actual native Q8 device route; no CPU fallback credit. Additional cost two graph encoders,8192 terminal bytes, one grouped serial quartet. No extra whole-token/prefill execution; same60/300 cap pending independent source/budget review. Combined quantization/projection boundary only; any unequal norm row remains an earlier mismatch. Strict oracle and production unchanged.

**Comment by anthony-chaudhary** [comment]
Projection diagnostic executed once: session82281 exit1, native Apple M3 Pro,1 run/0 skips. All4 normalized draft rows now match host bits exactly on frozen #12889 candidate. Token7 mixed-QKV output still differs at channel0: graph bf7e14f1 versus serial bf7e14f0. Serial route witnessed completed native q8-gemv-group with zero fallback; graph3encoders/1readback/12288bytes/0intermediate. This localizes the first measured remaining mismatch to combined Q8 activation quantization/projection, not yet one of those operations individually. Original strict assertions remain RED. Source55cb unchanged before/after; CPU/HIL slot and lease released. Receipt/raw/readable logs in managed559b86d76146/_scratch/issue-12853-projection-hil/. No retry or production edit. Next: independently scope quantization-versus-projection isolation.

**Comment by anthony-chaudhary** [comment]
Next one-file H/D/S diagnostic: upload exact per-row serial quantizeVecQ8 codes/scales; compare host-quantized graph GEMM H, existing device-quantized graph GEMM D, and serial grouped GEMV S on the same weight/input. Preserve all norm and strict checks. H differs S establishes GEMM/GEMV output drift with fixed serial quantization; H equals D does not prove code equality. Extra one encoder,8192 readback bytes (total20480),1152 upload bytes (total6272); no extra fence/read count or whole-token work. Source review and separate60/300 HIL grant required.

**Comment by anthony-chaudhary** [comment]
H/D/S physical diagnostic completed once: session58110 exit1,1 run/0 skips, native M3 Pro. All4 norms exact. H(host-quantized graph GEMM), D(device-quantized graph GEMM), S(serial grouped GEMV) all differ: channel0 H=bf7e14ef,D=bf7e14f1,S=bf7e14f0. H-vsS establishes projection-route drift with serial quantization fixed; H-vsD additionally shows activation-quantization/input-buffer-source sensitivity, without direct code/scale attribution. Required4encoders/20480read/6272upload and native S completion passed. Strict RED preserved; fullmanifest893ed unchanged, slot/lease released. Receipt and logs: managed559b86d76146/_scratch/issue-12853-hds-hil/. Next: independently scope direct activation code/scale evidence alongside the witnessed GEMM/GEMV difference.

**Comment by anthony-chaudhary** [comment]
Direct code/scale follow-on #12914 now has a physical observation independently checked against raw output: https://github.com/anthony-chaudhary/fak/issues/12914#issuecomment-5647991601

One native Apple M3 Pro run, session56092 exit1, 1 test/0 skips. All1024 codes match; 3/32 scale bit patterns differ (first row1/token5/block5: host3cd63ec4/device3cd63ec3). Every scale was compared; hashes/counts and first difference were persisted, not full arrays or the remaining two locations. All four normalized rows and the previous H/D/S hashes are unchanged. Lifecycle checks pass and the separate1152-byte diagnostic transfer leaves production20480-read/6272-upload/4-encoder accounting intact.

This is direct evidence of scale-bit drift in this exact fixture. H-versus-S remains separate evidence of projection-route drift under fixed host quantization. Neither proves a sole cause. Original strict parity remains RED, including logits/state/hidden/continuation.

Patch7eac3dd0/full-source-manifestaa7283 unchanged across all1702 checked entries. Physical receipt SHA2561a33c28ca1e0cdda0fef5709d11b2d64a10490968d0548f2b6008b928e59c598; raw JSONL6070ac396d89d9e124ed0cedfe7a48c74f155764367b9fac80cb5f341b5ce3b8. Artifacts preserved in managed model-165612fe8f4e/_scratch/issue-12914-hil. Owned processes absent and lease released; no repeat run.

Next: source-only review of scale formation/compiler behavior before a separately scoped arithmetic candidate; retain the independent projection-route investigation.

**Comment by anthony-chaudhary** [comment]
One physical #12919 candidate witness completed: actual executor5765 exit1, native Apple M3 Pro, one test/zero skips. Patch04207a/q4kf7c3/full36122/testd682 unchanged across all1702 source entries. Owned group12273 and recorded descendants absent at terminal; canonical lease released; no retry.

Direct result on this exact P4/H256 fixture:
- All1024 codes match exactly; all32 scale bit patterns now match exactly (previous diagnostic had three scale mismatches). Common code SHA256cdfecda26215f8bef90149c1afb481689f2723facdbbeb7237a0f2f592e7e588; common scale SHA25600f8cb88a3a83b03c56be3fffbd306054c226dc350491462a33381bd081e04fb.
- All four normalized rows remain exact. Token7 layer0 mixed-QKV H and D now match: SHA256f2b0f3b38465ff8a36ce4ede53656d3c242f5860c95225461efc8285c5d19a3e.
- H-versus-S remains different. Serial S remains SHA256cf28fa88ba3dbc73633df3c321ed8127753dfb02f84d31f300c02b0786f7deb0; channel0 H/D bf7e14ef versus S bf7e14f0.
- Lifecycle rejection and owned-copy checks pass. Separate diagnostic1152bytes; production one read20480bytes/upload6272bytes/four encoders, total observed21632readbytes.
- Full strict overall parity remains RED for logits/GDN/KV/hidden/continuation. No oracle or tolerance changed. The whole-quantizer compilation change supports only this fixture; it does not identify a sole compiler transformation or establish general exactness/performance.

Build emitted the expected fastMathEnabled deprecation warning and duplicate-lobjc linker warning; initialization and native execution succeeded. Test1.27s/package1.794s and supervised61.480s are operational records only. Independently observed peer landing PID13018 started12:11:31 PDT, overlapping the candidate window12:10:52–12:11:53 and test12:11:51.546832–12:11:52.813724. Coordinator also reported build13166; that PID was absent at post-run census, so its exact start is not independently witnessed. Peer untouched; no performance timing claim.

Preserved artifacts: managed `fak-worker-wt-metalgemm-fa700762879c/_scratch/issue-12919-hil/{receipt.json,supervision.json,test.jsonl,test.log,before-check.json,after-check.json}`. Raw JSONL SHA256ffa347447a0a3742e055919bea9810807a395425220b6a42f1044579fc72b24b. Receipt explicitly records competing work; source candidate remains frozen and unlanded.

Next: independent source-only scoping of Q8 graph projection reduction/accumulation against native grouped GEMV, preserving P4 batching and the original strict oracle. No additional run or kernel/algorithm edit before candidate review. Parent #12349 remains open; #12853 and #12914 evidence remains provenance-distinct.

**Comment by anthony-chaudhary** [comment]
One granted functional/bitwise invocation completed for #12920: actual executor8090 exit1, native Apple M3 Pro, two serialized packages, seven named tests run, zero skips. Six pass; original full strict parity remains RED.

PASS:
- TestProjectionGraphQ8P4MatchesSerialGEMV: all four rows/all37 outputs at in1280 (40 blocks), both graph entrypoints bitwise equal to native P1.
- TestProjectionGraphMixedQuantizedSingleFence (non-P4 graph).
- TestIssue8833Q8CallerOwnedBridge.
- TestMixedQ4KQ8Observation.
- TestMetalQ8GemvMatchesCPU (including5120x5120).
- TestMetalQ8GemvGroupMatchesSingle.

Original TestMetalMTPP4StrictSerialParity executes and remains FAIL. Its direct intermediate observations now show all1024 codes and all32 scale bit patterns exact, all four norm rows exact, and H=D=S for token7 layer0 mixed QKV, common SHA256cf28fa88ba3dbc73633df3c321ed8127753dfb02f84d31f300c02b0786f7deb0. Lifecycle checks and separate1152-byte diagnostic accounting remain valid; production20480read/6272upload/four encoders unchanged.

Remaining strict mismatch logs: verification logits rows1 and3, committed boundary, layer3 KRaw/KRoPE/V, and target-hidden history at commit/continuation. Earlier layer0–2 GDN and continuation-logit mismatch logs no longer appear. This does not establish full model parity; frozen source does not contain the separately owned hidden-capture overlays.

All1702 source entries remain unchanged at fullmanifestab896f4bd0b5051b607da2d7f578cef57fd9b5c7555e53af863e6f2be8fdccf8. Owned PID/PGID69547 and recorded descendants absent at terminal; no TERM/KILL/timeout/device error. Lease released. Resident80431 retains its original incarnation and was untouched. Functional contention policy allowed concurrent CPU builders and possiblyactive resident80431; elapsed204.527s is operational only, with no performance inference.

Preserved artifacts in managed fak-worker-wt-metalgemm-be633f987f73/_scratch/issue-12920-hil/{receipt.json,test.jsonl,test.log,supervision.json,before-check.json,after-check.json}. Receipt SHA256304d3b2a359ea48ab1a24cc398d07acaeb4276f782971d2525127677cfc062c6; rawJSONL4fea76bdc9fb293cb301d240a8b22540e89f905df389e2eb4ee98ca9b8ae899e. Source remains frozen/unlanded, one invocation only, no retry.

Next: independent raw-evidence review, then scope the remaining layer3 attention/logit discrepancy separately from the already tracked hidden-history work. Parent #12349 remains open; no new edit or run is implied.
